docs/reference/operator_parameters.md
<h1>Configuration parameters</h1>

There are two mutually-exclusive methods to set the Postgres Operator
configuration.

* ConfigMaps-based, the legacy one.  The configuration is supplied in a
  key-value configmap, defined by the `CONFIG_MAP_NAME` environment variable.
  Non-scalar values, i.e. lists or maps, are encoded in the value strings using
  the comma-based syntax for lists and coma-separated `key:value` syntax for
  maps. String values containing ':' should be enclosed in quotes. The
  configuration is flat, parameter group names below are not reflected in the
  configuration structure. There is an
  [example](../../manifests/configmap.yaml)

* CRD-based configuration. The configuration is stored in a custom YAML
  manifest. The manifest is an instance of the custom resource definition (CRD)
  called `OperatorConfiguration`. The operator registers this CRD during the
  start and uses it for configuration if the [operator deployment manifest](../../manifests/postgres-operator.yaml#L36)
  sets the `POSTGRES_OPERATOR_CONFIGURATION_OBJECT` env variable to a non-empty
  value. The variable should point to the `postgresql-operator-configuration`
  object in the operator's namespace.

  The CRD-based configuration is a regular YAML document; non-scalar keys are
  simply represented in the usual YAML way. There are no default values built-in
  in the operator, each parameter that is not supplied in the configuration
  receives an empty value. In order to create your own configuration just copy
  the [default one](../../manifests/postgresql-operator-default-configuration.yaml)
  and change it.

  To test the CRD-based configuration locally, use the following
  ```bash
  kubectl create -f manifests/operatorconfiguration.crd.yaml # registers the CRD
  kubectl create -f manifests/postgresql-operator-default-configuration.yaml

  kubectl create -f manifests/operator-service-account-rbac.yaml
  kubectl create -f manifests/postgres-operator.yaml # set the env var as mentioned above

  kubectl get operatorconfigurations postgresql-operator-default-configuration -o yaml
  ```

The CRD-based configuration is more powerful than the one based on ConfigMaps
and should be used unless there is a compatibility requirement to use an already
existing configuration. Even in that case, it should be rather straightforward
to convert the ConfigMap-based configuration into the CRD-based one and restart
the operator. The ConfigMap-based configuration will be deprecated and
subsequently removed in future releases.

Note that for the CRD-based configuration groups of configuration options below
correspond to the non-leaf keys in the target YAML (i.e. for the Kubernetes
resources the key is `kubernetes`). The key is mentioned alongside the group
description. The ConfigMap-based configuration is flat and does not allow
non-leaf keys.

Since in the CRD-based case the operator needs to create a CRD first, which is
controlled by the `resource_check_interval` and `resource_check_timeout`
parameters, those parameters have no effect and are replaced by the
`CRD_READY_WAIT_INTERVAL` and `CRD_READY_WAIT_TIMEOUT` environment variables.
They will be deprecated and removed in the future.

For the configmap configuration, the [default parameter values](../../pkg/util/config/config.go#L14)
mentioned here are likely to be overwritten in your local operator installation
via your local version of the operator configmap. In the case you use the
operator CRD, all the CRD defaults are provided in the
[operator's default configuration manifest](../../manifests/postgresql-operator-default-configuration.yaml)

Variable names are underscore-separated words.


## General

Those are top-level keys, containing both leaf keys and groups.

* **enable_crd_validation**
  toggles if the operator will create or update CRDs with
  [OpenAPI v3 schema validation](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#validation)
  The default is `true`.

* **enable_lazy_spilo_upgrade**
  Instruct operator to update only the statefulsets with new images (Spilo and InitContainers) without immediately doing the rolling update. The assumption is pods will be re-started later with new images, for example due to the node rotation.
  The default is `false`.

* **enable_pgversion_env_var**
  With newer versions of Spilo, it is preferable to use `PGVERSION` pod environment variable instead of the setting `postgresql.bin_dir` in the `SPILO_CONFIGURATION` env variable. When this option is true, the operator sets `PGVERSION` and omits `postgresql.bin_dir` from  `SPILO_CONFIGURATION`. When false, the `postgresql.bin_dir` is set. This setting takes precedence over `PGVERSION`; see PR 222 in Spilo. The default is `true`.

* **enable_spilo_wal_path_compat**
  enables backwards compatible path between Spilo 12 and Spilo 13 images. The default is `false`.

* **etcd_host**
  Etcd connection string for Patroni defined as `host:port`. Not required when
  Patroni native Kubernetes support is used. The default is empty (use
  Kubernetes-native DCS).

* **kubernetes_use_configmaps**
  Select if setup uses endpoints (default), or configmaps to manage leader when
  DCS is kubernetes (not etcd or similar). In OpenShift it is not possible to
  use endpoints option, and configmaps is required. By default,
  `kubernetes_use_configmaps: false`, meaning endpoints will be used.

* **docker_image**
  Spilo Docker image for Postgres instances. For production, don't rely on the
  default image, as it might be not the most up-to-date one. Instead, build
  your own Spilo image from the [github
  repository](https://github.com/zalando/spilo).

* **sidecar_docker_images**
  *deprecated*: use **sidecars** instead. A map of sidecar names to Docker
  images to run with Spilo. In case of the name conflict with the definition in
  the cluster manifest the cluster-specific one is preferred.

* **sidecars**
  a list of sidecars to run with Spilo, for any cluster (i.e. globally defined
  sidecars). Each item in the list is of type
  [Container](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#container-v1-core).
  Globally defined sidecars can be overwritten by specifying a sidecar in the
  Postgres manifest with the same name.
  Note: This field is not part of the schema validation. If the container
  specification is invalid, then the operator fails to create the statefulset.

* **enable_shm_volume**
  Instruct operator to start any new database pod without limitations on shm
  memory. If this option is enabled, to the target database pod will be mounted
  a new tmpfs volume to remove shm memory limitation (see e.g. the
  [docker issue](https://github.com/docker-library/postgres/issues/416)).
  This option is global for an operator object, and can be overwritten by
  `enableShmVolume` parameter from Postgres manifest. The default is `true`.

* **workers**
  number of working routines the operator spawns to process requests to
  create/update/delete/sync clusters concurrently. The default is `8`.

* **max_instances**
  operator will cap the number of instances in any managed Postgres cluster up
  to the value of this parameter. When `-1` is specified, no limits are applied.
  The default is `-1`.

* **min_instances**
  operator will run at least the number of instances for any given Postgres
  cluster equal to the value of this parameter. Standby clusters can still run
  with `numberOfInstances: 1` as this is the [recommended setup](../user.md#setting-up-a-standby-cluster).
  When `-1` is specified for `min_instances`, no limits are applied. The default
  is `-1`.

* **resync_period**
  period between consecutive sync requests. The default is `30m`.

* **repair_period**
  period between consecutive repair requests. The default is `5m`.

* **set_memory_request_to_limit**
  Set `memory_request` to `memory_limit` for all Postgres clusters (the default
  value is also increased). This prevents certain cases of memory overcommitment
  at the cost of overprovisioning memory and potential scheduling problems for
  containers with high memory limits due to the lack of memory on Kubernetes
  cluster nodes. This affects all containers created by the operator (Postgres,
  Scalyr sidecar, and other sidecars except **sidecars** defined in the operator
  configuration); to set resources for the operator's own container, change the
  [operator deployment manually](../../manifests/postgres-operator.yaml#L20).
  The default is `false`.

* **read_only_root_filesystem**
  This sets readOnlyRootFilesystem for the Postgres pods.
  The default is `false`.

## Postgres users

Parameters describing Postgres users. In a CRD-configuration, they are grouped
under the `users` key.

* **super_username**
  Postgres `superuser` name to be created by `initdb`. The default is
  `postgres`.

* **replication_username**
  Postgres username used for replication between instances. The default is
  `standby`.

## Major version upgrades

Parameters configuring automatic major version upgrades. In a
CRD-configuration, they are grouped under the `major_version_upgrade` key.

* **major_version_upgrade_mode**
  Postgres Operator supports [in-place major version upgrade](../administrator.md#in-place-major-version-upgrade)
  with three different modes:
  `"off"` = no upgrade by the operator,
  `"manual"` = manifest triggers action,
  `"full"` = manifest and minimal version violation trigger upgrade.
  Note, that with all three modes increasing the version in the manifest will
  trigger a rolling update of the pods. The default is `"off"`.

* **minimal_major_version**
  The minimal Postgres major version that will not automatically be upgraded
  when `major_version_upgrade_mode` is set to `"full"`. The default is `"9.5"`.

* **target_major_version**
  The target Postgres major version when upgrading clusters automatically
  which violate the configured allowed `minimal_major_version` when
  `major_version_upgrade_mode` is set to `"full"`. The default is `"13"`.

## Kubernetes resources

Parameters to configure cluster-related Kubernetes objects created by the
operator, as well as some timeouts associated with them. In a CRD-based
configuration they are grouped under the `kubernetes` key.

* **pod_service_account_name**
  service account used by Patroni running on individual Pods to communicate
  with the operator. Required even if native Kubernetes support in Patroni is
  not used, because Patroni keeps pod labels in sync with the instance role.
  The default is `postgres-pod`.

* **pod_service_account_definition**
  On Postgres cluster creation the operator tries to create the service account
  for the Postgres pods if it does not exist in the namespace. The internal
  default service account definition (defines only the name) can be overwritten
  with this parameter. Make sure to provide a valid YAML or JSON string. The
  default is empty.

* **pod_service_account_role_binding_definition**
  This definition must bind the pod service account to a role with permission
  sufficient for the pods to start and for Patroni to access K8s endpoints;
  service account on its own lacks any such rights starting with K8s v1.8. If
  not explicitly defined by the user, a simple definition that binds the
  account to the 'postgres-pod' [cluster role](../../manifests/operator-service-account-rbac.yaml#L198)
  will be used. The default is empty.

* **pod_terminate_grace_period**
  Postgres pods are [terminated forcefully](https://kubernetes.io/docs/concepts/workloads/pods/pod/#termination-of-pods)
  after this timeout. The default is `5m`.

* **custom_pod_annotations**
  This key/value map provides a list of annotations that get attached to each pod
  of a database created by the operator. If the annotation key is also provided
  by the database definition, the database definition value is used.

* **delete_annotation_date_key**
  key name for annotation that compares manifest value with current date in the
  YYYY-MM-DD format. Allowed pattern: `'([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]'`.
  The default is empty which also disables this delete protection check.

* **delete_annotation_name_key**
  key name for annotation that compares manifest value with Postgres cluster name.
  Allowed pattern: `'([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]'`. The default is
  empty which also disables this delete protection check.

* **downscaler_annotations**
  An array of annotations that should be passed from Postgres CRD on to the
  statefulset and, if exists, to the connection pooler deployment as well.
  Regular expressions like `downscaler/*` etc. are also accepted. Can be used
  with [kube-downscaler](https://github.com/hjacobs/kube-downscaler).

* **watched_namespace**
  The operator watches for Postgres objects in the given namespace. If not
  specified, the value is taken from the operator namespace. A special `*`
  value makes it watch all namespaces. The default is empty (watch the operator
  pod namespace).

* **pdb_name_format**
  defines the template for PDB (Pod Disruption Budget) names created by the
  operator. The default is `postgres-{cluster}-pdb`, where `{cluster}` is
  replaced by the cluster name. Only the `{cluster}` placeholders is allowed in
  the template.

* **enable_pod_disruption_budget**
  PDB is enabled by default to protect the cluster from voluntarily disruptions
  and hence unwanted DB downtime. However, on some cloud providers it could be
  necessary to temporarily disabled it, e.g. for node updates. See
  [admin docs](../administrator.md#pod-disruption-budget) for more information.
  Default is true.

* **enable_cross_namespace_secrets**
  To allow secrets in a different namespace other than the Postgres cluster
  namespace. Once enabled, specify the namespace in the user name under the
  `users` section in the form `{namespace}.{username}`. The operator will then
  create the user secret in that namespace. The part after the first `.` is
  considered to be the user name. The default is `false`.

* **enable_init_containers**
  global option to allow for creating init containers in the cluster manifest to
  run actions before Spilo is started. Default is true.

* **enable_sidecars**
  global option to allow for creating sidecar containers in the cluster manifest
  to run alongside Spilo on the same pod. Globally defined sidecars are always
  enabled. Default is true.

* **secret_name_template**
  a template for the name of the database user secrets generated by the
  operator. `{namespace}` is replaced with name of the namespace if
  `enable_cross_namespace_secret` is set, otherwise the
  secret is in cluster's namespace. `{username}` is replaced with name of the
  secret, `{cluster}` with the name of the cluster, `{tprkind}` with the kind
  of CRD (formerly known as TPR) and `{tprgroup}` with the group of the CRD.
  No other placeholders are allowed. The default is
  `{namespace}.{username}.{cluster}.credentials.{tprkind}.{tprgroup}`.

* **cluster_domain**
  defines the default DNS domain for the kubernetes cluster the operator is
  running in. The default is `cluster.local`. Used by the operator to connect
  to the Postgres clusters after creation.

* **oauth_token_secret_name**
  namespaced name of the secret containing the `OAuth2` token to pass to the
  teams API. The default is `postgresql-operator`.

* **infrastructure_roles_secret_name**
  *deprecated*: namespaced name of the secret containing infrastructure roles
  with user names, passwords and role membership.

* **infrastructure_roles_secrets**
  array of infrastructure role definitions which reference existing secrets
  and specify the key names from which user name, password and role membership
  are extracted. For the ConfigMap this has to be a string which allows
  referencing only one infrastructure roles secret. The default is empty.

* **inherited_annotations**
  list of annotation keys that can be inherited from the cluster manifest, and
  added to each child objects  (`Deployment`, `StatefulSet`, `Pod`, `PDB` and
  `Services`) created by the operator incl. the ones from the connection
  pooler deployment. The default is empty.

* **pod_role_label**
  name of the label assigned to the Postgres pods (and services/endpoints) by
  the operator. The default is `spilo-role`.

* **cluster_labels**
  list of `name:value` pairs for additional labels assigned to the cluster
  objects. The default is `application:spilo`.

* **inherited_labels**
  list of label keys that can be inherited from the cluster manifest, and
  added to each child objects (`Deployment`, `StatefulSet`, `Pod`, `PVCs`,
  `PDB`, `Service`, `Endpoints` and `Secrets`) created by the operator.
  Typical use case is to dynamically pass labels that are specific to a
  given Postgres cluster, in order to implement `NetworkPolicy`. The default
  is empty.

* **cluster_name_label**
  name of the label assigned to Kubernetes objects created by the operator
  that indicates which cluster a given object belongs to. The default is
  `cluster-name`.

* **node_readiness_label**
  a set of labels that a running and active node should possess to be
  considered `ready`. The operator uses values of those labels to detect the
  start of the Kubernetes cluster upgrade procedure and move master pods off
  the nodes to be decommissioned. When the set is not empty, the operator also
  assigns the `Affinity` clause to the Postgres pods to be scheduled only on
  `ready` nodes. The default is empty.

* **toleration**
  a dictionary that should contain `key`, `operator`, `value` and
  `effect` keys. In that case, the operator defines a pod toleration
  according to the values of those keys. See [kubernetes documentation](https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/)
  for details on taints and tolerations. The default is empty.

* **pod_environment_configmap**
  namespaced name of the ConfigMap with environment variables to populate on
  every pod. Right now this ConfigMap is searched in the namespace of the
  Postgres cluster. All variables from that ConfigMap are injected to the pod's
  environment, on conflicts they are overridden by the environment variables
  generated by the operator. The default is empty.

* **pod_priority_class_name**

pkg/util/config/config.go
package config

import (
	"encoding/json"
	"strings"
	"time"

	"fmt"

	"github.com/zalando/postgres-operator/pkg/spec"
	"github.com/zalando/postgres-operator/pkg/util/constants"
	v1 "k8s.io/api/core/v1"
)

// CRD describes CustomResourceDefinition specific configuration parameters
type CRD struct {
	ReadyWaitInterval   time.Duration `name:"ready_wait_interval" default:"4s"`
	ReadyWaitTimeout    time.Duration `name:"ready_wait_timeout" default:"30s"`
	ResyncPeriod        time.Duration `name:"resync_period" default:"30m"`
	RepairPeriod        time.Duration `name:"repair_period" default:"5m"`
	EnableCRDValidation *bool         `name:"enable_crd_validation" default:"true"`
}

// Resources describes kubernetes resource specific configuration parameters
type Resources struct {
	ResourceCheckInterval         time.Duration       `name:"resource_check_interval" default:"3s"`
	ResourceCheckTimeout          time.Duration       `name:"resource_check_timeout" default:"10m"`
	PodLabelWaitTimeout           time.Duration       `name:"pod_label_wait_timeout" default:"10m"`
	PodDeletionWaitTimeout        time.Duration       `name:"pod_deletion_wait_timeout" default:"10m"`
	PodTerminateGracePeriod       time.Duration       `name:"pod_terminate_grace_period" default:"5m"`
	SpiloRunAsUser                *int64              `name:"spilo_runasuser"`
	SpiloRunAsGroup               *int64              `name:"spilo_runasgroup"`
	SpiloFSGroup                  *int64              `name:"spilo_fsgroup"`
	PodPriorityClassName          string              `name:"pod_priority_class_name"`
	ReadOnlyRootFilesystem        *bool               `name:"read_only_root_filesystem" default:"false"`
	ClusterDomain                 string              `name:"cluster_domain" default:"cluster.local"`
	SpiloPrivileged               bool                `name:"spilo_privileged" default:"false"`
	SpiloAllowPrivilegeEscalation *bool               `name:"spilo_allow_privilege_escalation" default:"true"`
	AdditionalPodCapabilities     []string            `name:"additional_pod_capabilities" default:""`
	ClusterLabels                 map[string]string   `name:"cluster_labels" default:"application:spilo"`
	InheritedLabels               []string            `name:"inherited_labels" default:""`
	InheritedAnnotations          []string            `name:"inherited_annotations" default:""`
	DownscalerAnnotations         []string            `name:"downscaler_annotations"`
	ClusterNameLabel              string              `name:"cluster_name_label" default:"cluster-name"`
	DeleteAnnotationDateKey       string              `name:"delete_annotation_date_key"`
	DeleteAnnotationNameKey       string              `name:"delete_annotation_name_key"`
	PodRoleLabel                  string              `name:"pod_role_label" default:"spilo-role"`
	PodToleration                 map[string]string   `name:"toleration" default:""`
	DefaultCPURequest             string              `name:"default_cpu_request" default:"100m"`
	DefaultMemoryRequest          string              `name:"default_memory_request" default:"100Mi"`
	DefaultCPULimit               string              `name:"default_cpu_limit" default:"1"`
	DefaultMemoryLimit            string              `name:"default_memory_limit" default:"500Mi"`
	MinCPULimit                   string              `name:"min_cpu_limit" default:"250m"`
	MinMemoryLimit                string              `name:"min_memory_limit" default:"250Mi"`
	PodEnvironmentConfigMap       spec.NamespacedName `name:"pod_environment_configmap"`
	PodEnvironmentSecret          string              `name:"pod_environment_secret"`
	NodeReadinessLabel            map[string]string   `name:"node_readiness_label" default:""`
	MaxInstances                  int32               `name:"max_instances" default:"-1"`
	MinInstances                  int32               `name:"min_instances" default:"-1"`
	ShmVolume                     *bool               `name:"enable_shm_volume" default:"true"`
}

type InfrastructureRole struct {
	// Name of a secret which describes the role, and optionally name of a
	// configmap with an extra information
	SecretName spec.NamespacedName

	UserKey     string
	PasswordKey string
	RoleKey     string

	DefaultUserValue string
	DefaultRoleValue string

	// This field point out the detailed yaml definition of the role, if exists
	Details string

	// Specify if a secret contains multiple fields in the following format:
	//
	// 	%(userkey)idx: ...
	// 	%(passwordkey)idx: ...
	// 	%(rolekey)idx: ...
	//
	// If it does, Name/Password/Role are interpreted not as unique field
	// names, but as a template.

	Template bool
}

// Auth describes authentication specific configuration parameters
type Auth struct {
	SecretNameTemplate            StringTemplate        `name:"secret_name_template" default:"{username}.{cluster}.credentials.{tprkind}.{tprgroup}"`
	PamRoleName                   string                `name:"pam_role_name" default:"zalandos"`
	PamConfiguration              string                `name:"pam_configuration" default:"https://info.example.com/oauth2/tokeninfo?access_token= uid realm=/employees"`
	TeamsAPIUrl                   string                `name:"teams_api_url" default:"https://teams.example.com/api/"`
	OAuthTokenSecretName          spec.NamespacedName   `name:"oauth_token_secret_name" default:"postgresql-operator"`
	InfrastructureRolesSecretName spec.NamespacedName   `name:"infrastructure_roles_secret_name"`
	InfrastructureRoles           []*InfrastructureRole `name:"-"`
	InfrastructureRolesDefs       string                `name:"infrastructure_roles_secrets"`
	SuperUsername                 string                `name:"super_username" default:"postgres"`
	ReplicationUsername           string                `name:"replication_username" default:"standby"`
}

// Scalyr holds the configuration for the Scalyr Agent sidecar for log shipping:
type Scalyr struct {
	ScalyrAPIKey        string `name:"scalyr_api_key" default:""`
	ScalyrImage         string `name:"scalyr_image" default:""`
	ScalyrServerURL     string `name:"scalyr_server_url" default:"https://upload.eu.scalyr.com"`
	ScalyrCPURequest    string `name:"scalyr_cpu_request" default:"100m"`
	ScalyrMemoryRequest string `name:"scalyr_memory_request" default:"50Mi"`
	ScalyrCPULimit      string `name:"scalyr_cpu_limit" default:"1"`
	ScalyrMemoryLimit   string `name:"scalyr_memory_limit" default:"500Mi"`
}

// LogicalBackup defines configuration for logical backup
type LogicalBackup struct {
	LogicalBackupSchedule                     string `name:"logical_backup_schedule" default:"30 00 * * *"`
	LogicalBackupDockerImage                  string `name:"logical_backup_docker_image" default:"registry.opensource.zalan.do/acid/logical-backup:v1.6.3"`
	LogicalBackupProvider                     string `name:"logical_backup_provider" default:"s3"`
	LogicalBackupS3Bucket                     string `name:"logical_backup_s3_bucket" default:""`
	LogicalBackupS3Region                     string `name:"logical_backup_s3_region" default:""`
	LogicalBackupS3Endpoint                   string `name:"logical_backup_s3_endpoint" default:""`
	LogicalBackupS3AccessKeyID                string `name:"logical_backup_s3_access_key_id" default:""`
	LogicalBackupS3SecretAccessKey            string `name:"logical_backup_s3_secret_access_key" default:""`
	LogicalBackupS3SSE                        string `name:"logical_backup_s3_sse" default:""`
	LogicalBackupGoogleApplicationCredentials string `name:"logical_backup_google_application_credentials" default:""`
	LogicalBackupJobPrefix                    string `name:"logical_backup_job_prefix" default:"logical-backup-"`
}

// Operator options for connection pooler
type ConnectionPooler struct {
	NumberOfInstances                    *int32 `name:"connection_pooler_number_of_instances" default:"2"`
	Schema                               string `name:"connection_pooler_schema" default:"pooler"`
	User                                 string `name:"connection_pooler_user" default:"pooler"`
	Image                                string `name:"connection_pooler_image" default:"registry.opensource.zalan.do/acid/pgbouncer"`
	Mode                                 string `name:"connection_pooler_mode" default:"transaction"`
	MaxDBConnections                     *int32 `name:"connection_pooler_max_db_connections" default:"60"`
	ConnectionPoolerDefaultCPURequest    string `name:"connection_pooler_default_cpu_request" default:"500m"`
	ConnectionPoolerDefaultMemoryRequest string `name:"connection_pooler_default_memory_request" default:"100Mi"`
	ConnectionPoolerDefaultCPULimit      string `name:"connection_pooler_default_cpu_limit" default:"1"`
	ConnectionPoolerDefaultMemoryLimit   string `name:"connection_pooler_default_memory_limit" default:"100Mi"`
}

// Config describes operator config
type Config struct {
	CRD
	Resources
	Auth
	Scalyr
	LogicalBackup
	ConnectionPooler

	WatchedNamespace        string            `name:"watched_namespace"` // special values: "*" means 'watch all namespaces', the empty string "" means 'watch a namespace where operator is deployed to'
	KubernetesUseConfigMaps bool              `name:"kubernetes_use_configmaps" default:"false"`
	EtcdHost                string            `name:"etcd_host" default:""` // special values: the empty string "" means Patroni will use K8s as a DCS
	DockerImage             string            `name:"docker_image" default:"registry.opensource.zalan.do/acid/spilo-13:2.0-p7"`
	SidecarImages           map[string]string `name:"sidecar_docker_images"` // deprecated in favour of SidecarContainers
	SidecarContainers       []v1.Container    `name:"sidecars"`
	PodServiceAccountName   string            `name:"pod_service_account_name" default:"postgres-pod"`
	// value of this string must be valid JSON or YAML; see initPodServiceAccount
	PodServiceAccountDefinition            string            `name:"pod_service_account_definition" default:""`
	PodServiceAccountRoleBindingDefinition string            `name:"pod_service_account_role_binding_definition" default:""`
	MasterPodMoveTimeout                   time.Duration     `name:"master_pod_move_timeout" default:"20m"`
	DbHostedZone                           string            `name:"db_hosted_zone" default:"db.example.com"`
	AWSRegion                              string            `name:"aws_region" default:"eu-central-1"`
	WALES3Bucket                           string            `name:"wal_s3_bucket"`
	LogS3Bucket                            string            `name:"log_s3_bucket"`
	KubeIAMRole                            string            `name:"kube_iam_role"`
	WALGSBucket                            string            `name:"wal_gs_bucket"`
	GCPCredentials                         string            `name:"gcp_credentials"`
	AdditionalSecretMount                  string            `name:"additional_secret_mount"`
	AdditionalSecretMountPath              string            `name:"additional_secret_mount_path" default:"/meta/credentials"`
	EnableEBSGp3Migration                  bool              `name:"enable_ebs_gp3_migration" default:"false"`
	EnableEBSGp3MigrationMaxSize           int64             `name:"enable_ebs_gp3_migration_max_size" default:"1000"`
	DebugLogging                           bool              `name:"debug_logging" default:"true"`
	EnableDBAccess                         bool              `name:"enable_database_access" default:"true"`
	EnableTeamsAPI                         bool              `name:"enable_teams_api" default:"true"`
	EnableTeamSuperuser                    bool              `name:"enable_team_superuser" default:"false"`
	TeamAdminRole                          string            `name:"team_admin_role" default:"admin"`
	RoleDeletionSuffix                     string            `name:"role_deletion_suffix" default:"_deleted"`
	EnableTeamMemberDeprecation            bool              `name:"enable_team_member_deprecation" default:"false"`
	EnableAdminRoleForUsers                bool              `name:"enable_admin_role_for_users" default:"true"`
	EnablePostgresTeamCRD                  bool              `name:"enable_postgres_team_crd" default:"false"`
	EnablePostgresTeamCRDSuperusers        bool              `name:"enable_postgres_team_crd_superusers" default:"false"`
	EnableMasterLoadBalancer               bool              `name:"enable_master_load_balancer" default:"true"`
	EnableReplicaLoadBalancer              bool              `name:"enable_replica_load_balancer" default:"false"`
	CustomServiceAnnotations               map[string]string `name:"custom_service_annotations"`
	CustomPodAnnotations                   map[string]string `name:"custom_pod_annotations"`
	EnablePodAntiAffinity                  bool              `name:"enable_pod_antiaffinity" default:"false"`
	PodAntiAffinityTopologyKey             string            `name:"pod_antiaffinity_topology_key" default:"kubernetes.io/hostname"`
	StorageResizeMode                      string            `name:"storage_resize_mode" default:"pvc"`
	EnableLoadBalancer                     *bool             `name:"enable_load_balancer"` // deprecated and kept for backward compatibility
	ExternalTrafficPolicy                  string            `name:"external_traffic_policy" default:"Cluster"`
	MasterDNSNameFormat                    StringTemplate    `name:"master_dns_name_format" default:"{cluster}.{team}.{hostedzone}"`
	ReplicaDNSNameFormat                   StringTemplate    `name:"replica_dns_name_format" default:"{cluster}-repl.{team}.{hostedzone}"`
	PDBNameFormat                          StringTemplate    `name:"pdb_name_format" default:"postgres-{cluster}-pdb"`
	EnablePodDisruptionBudget              *bool             `name:"enable_pod_disruption_budget" default:"true"`
	EnableInitContainers                   *bool             `name:"enable_init_containers" default:"true"`
	EnableSidecars                         *bool             `name:"enable_sidecars" default:"true"`
	Workers                                uint32            `name:"workers" default:"8"`
	APIPort                                int               `name:"api_port" default:"8080"`
	RingLogLines                           int               `name:"ring_log_lines" default:"100"`
	ClusterHistoryEntries                  int               `name:"cluster_history_entries" default:"1000"`
	TeamAPIRoleConfiguration               map[string]string `name:"team_api_role_configuration" default:"log_statement:all"`
	PodTerminateGracePeriod                time.Duration     `name:"pod_terminate_grace_period" default:"5m"`
	PodManagementPolicy                    string            `name:"pod_management_policy" default:"ordered_ready"`
	ProtectedRoles                         []string          `name:"protected_role_names" default:"admin"`
	PostgresSuperuserTeams                 []string          `name:"postgres_superuser_teams" default:""`
	SetMemoryRequestToLimit                bool              `name:"set_memory_request_to_limit" default:"false"`
	EnableLazySpiloUpgrade                 bool              `name:"enable_lazy_spilo_upgrade" default:"false"`
	EnableCrossNamespaceSecret             bool              `name:"enable_cross_namespace_secret" default:"false"`
	EnablePgVersionEnvVar                  bool              `name:"enable_pgversion_env_var" default:"true"`
	EnableSpiloWalPathCompat               bool              `name:"enable_spilo_wal_path_compat" default:"false"`
	MajorVersionUpgradeMode                string            `name:"major_version_upgrade_mode" default:"off"`
	MinimalMajorVersion                    string            `name:"minimal_major_version" default:"9.5"`
	TargetMajorVersion                     string            `name:"target_major_version" default:"13"`
}

// MustMarshal marshals the config or panics
func (c Config) MustMarshal() string {
	b, err := json.MarshalIndent(c, "", "   ")
	if err != nil {
		panic(err)
	}

	return string(b)
}

// NewFromMap creates Config from the map
func NewFromMap(m map[string]string) *Config {
	cfg := Config{}
	fields, _ := structFields(&cfg)

	for _, structField := range fields {
		key := strings.ToLower(structField.Name)
		value, ok := m[key]
		if !ok && structField.Default != "" {
			value = structField.Default
		}

		if value == "" {
			continue
		}
		err := processField(value, structField.Field)
		if err != nil {
			panic(err)
		}
	}
	if err := validate(&cfg); err != nil {
		panic(err)
	}

	return &cfg
}

// Copy creates a copy of the config
func Copy(c *Config) Config {
	cfg := *c

	cfg.ClusterLabels = make(map[string]string, len(c.ClusterLabels))
	for k, v := range c.ClusterLabels {
		cfg.ClusterLabels[k] = v
	}

	return cfg
}

func validate(cfg *Config) (err error) {
	if cfg.MinInstances > 0 && cfg.MaxInstances > 0 && cfg.MinInstances > cfg.MaxInstances {
		err = fmt.Errorf("minimum number of instances %d is set higher than the maximum number %d",
			cfg.MinInstances, cfg.MaxInstances)
	}
	if cfg.Workers == 0 {
		err = fmt.Errorf("number of workers should be higher than 0")
	}

	if *cfg.ConnectionPooler.NumberOfInstances < constants.ConnectionPoolerMinInstances {
		msg := "number of connection pooler instances should be higher than %d"
		err = fmt.Errorf(msg, constants.ConnectionPoolerMinInstances)
	}

	if cfg.ConnectionPooler.User == cfg.SuperUsername {
		msg := "Connection pool user is not allowed to be the same as super user, username: %s"
		err = fmt.Errorf(msg, cfg.ConnectionPooler.User)
	}

	return
}
